File: leetcode_java/java/com/goinhn/leetcode/_1_100/Sol94.java
package com.goinhn.leetcode._1_100;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * @author goinhn
 * @date 2021/11/20
 */
public class Sol94 {

    private static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    /**
     * 递归实现
     * <p>
     * time:O(n)
     * space:O(n)
     */
    private static class Solution1 {
        public List<Integer> inorderTraversal(TreeNode root) {
            List<Integer> list = new ArrayList<>();
            setVal(root, list);
            return list;
        }

        private void setVal(TreeNode root, List<Integer> list) {
            if (root != null) {
                setVal(root.left, list);
                list.add(root.val);
                setVal(root.right, list);
            }
        }
    }

    /**
     * 利用栈进行迭代
     * <p>
     * time:O(n)
     * space:O(n)
     */
    private static class Solution2 {
        public List<Integer> inorderTraversal(TreeNode root) {
            List<Integer> list = new ArrayList<>();
            Stack<TreeNode> stack = new Stack<>();
            while (!stack.empty() || root != null) {
                while (root != null) {
                    stack.push(root);
                    root = root.left;
                }
                root = stack.pop();
                list.add(root.val);
                root = root.right;
            }
            return list;
        }
    }
}

File: leetcode_go/question/_1_100/_94/sol_94.go
// Package _94 二叉树的中序遍历
// @author goinhn
// @data 2022/8/19
package _94

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 递归遍历
// time:O(n)
// space:O(n)
func inorderTraversal1(root *TreeNode) (res []int) {
	var inOrderFunc func(root *TreeNode)
	inOrderFunc = func(root *TreeNode) {
		if root == nil {
			return
		}

		inOrderFunc(root.Left)
		res = append(res, root.Val)
		inOrderFunc(root.Right)
	}

	inOrderFunc(root)
	return res
}

// 递归遍历
// time:O(n)
// space:O(n)
func inorderTraversal2(root *TreeNode) []int {
	var res []int
	dfs(root, &res)

	return res
}

func dfs(root *TreeNode, res *[]int) {
	if root == nil {
		return
	}

	dfs(root.Left, res)
	*res = append(*res, root.Val)
	dfs(root.Right, res)
}

// 迭代遍历
// time:O(n)
// space:O(n)
func inorderTraversal3(root *TreeNode) []int {
	var res []int
	var stack []*TreeNode
	for root != nil || len(stack) > 0 {
		for root != nil {
			stack = append(stack, root)
			root = root.Left
		}

		root = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		res = append(res, root.Val)
		root = root.Right
	}

	return res
}
